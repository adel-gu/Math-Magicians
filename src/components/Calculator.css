button {
  border: none;
  padding: 15px 10px;
  background-color: aliceblue;
}

.calculator {
  width: 450px;
  max-width: 90%;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  gap: 1px;
  box-shadow: 10px 8px 28px -7px rgba(0, 0, 0, 0.15);
}

.screen {
  background: var(--clr-screen);

  grid-column: span 4;
  padding: 5px;
  text-align: right;
}

.numberZero {
  grid-column: span 2;
}

.orang {
  background-color: var(--clr-orange);
  color: white;
}

button:hover {
  cursor: pointer;
  background: rgba(0, 0, 0, 0.1);
}

button.orang:hover {
  background-color: rgba(245, 145, 62, 0.8);
}
