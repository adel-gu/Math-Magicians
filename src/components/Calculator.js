import '../App.css';

const Screen = (props) => {
  return <span className='screen'>{props.result}</span>;
};

const Oprand = ({ data, className }) => {
  return <button className={className}>{data}</button>;
};

const Number = ({ value, className }) => {
  return <button className={className}>{value}</button>;
};

// Calculator

const Calculator = () => {
  return (
    <div className='calculator'>
      <Screen result='0' />
      {/* First row */}
      <Oprand data='AC' />
      <Oprand data='+/-' />
      <Oprand data='%' />
      <Oprand data='+' className='orang' />
      {/* Seccond row */}
      <Number value='7' />
      <Number value='8' />
      <Number value='9' />
      <Oprand data='*' className='orang' />
      {/* 3rd row */}
      <Number value='4' />
      <Number value='5' />
      <Number value='6' />
      <Oprand data='-' className='orang' />
      {/* 4th row */}
      <Number value='1' />
      <Number value='2' />
      <Number value='3' />
      <Oprand data='+' className='orang' />
      {/* 5th row */}
      <Number value='0' className='numberZero' />
      <Number value='.' />
      <Oprand data='=' className='orang' />
    </div>
  );
};

export default Calculator;
