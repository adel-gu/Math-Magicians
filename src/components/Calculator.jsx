import { useState } from 'react';
import PropTypes from 'prop-types';
import { Component } from 'react';
import calculate from "../logic/calculate";

// Sub Components => {Screen, oprands, Numbers}
const Screen = ({ result }) => <span className="screen">{result}</span>;
Screen.propTypes = { result: PropTypes.string.isRequired };

const Oprand = ({ data, className, onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {data}
  </button>
);
Oprand.propTypes = { data: PropTypes.string.isRequired };
Oprand.propTypes = { className: PropTypes.string };
Oprand.propTypes = { onclick: PropTypes.func.isRequired };

const Number = ({ value, className , onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {value}
  </button>
);
Number.propTypes = { value: PropTypes.string.isRequired };
Number.propTypes = { className: PropTypes.string};
Number.propTypes = { onclick: PropTypes.func.isRequired};

// Calculator Component
const Calculator = () => {
  const [calculator, setCalculator] = useState({
    total: '0',
    next: null,
    operation: null
  });
  // Numbers
  const numValues = [0,1,2,3,4,5,6,7,8,9,'.'];

  // Calculation method
  const handlCalculation = (buttonName) => {
    const result = calculate(calculator, buttonName.toString());
    setCalculator({...result})
  }

  // Render the calculated result on the screen component.
  const renderScreen = () => {
    let total = calculator.total;
    let next = calculator.next;
    let operation = calculator.operation;

    let screen = `${total || ''}${operation || ''}${next || ''}`;
    // If all the properties are null 
    if (!(total || next || operation)) {
      screen = "0";
    }
    
    return (
      <Screen result={screen}/>
    )
  }

  // Render the numbers component
  const renderNumber = (i, className) => {
    return (
      <Number 
        value={numValues[i]} 
        className={className}
        onclick= {() => handlCalculation(numValues[i])}
      />
    )
  }

  // Render Oprands
  const renderOprand = (data, className) => {
    return (
      <Oprand 
        data={data} 
        className={className}
        onclick= {() => handlCalculation(data)}
      />
    )
  }

  // Render all calculator components
  return (
    <div className="calculator">
      {renderScreen()}
      {/* First row */}
      {renderOprand('AC')}
      {renderOprand('+/-')}
      {renderOprand('%')}
      {renderOprand('รท', "orang")}
      {/* Seccond row */}
      {renderNumber(7)}
      {renderNumber(8)}
      {renderNumber(9)}
      {renderOprand('x', "orang")}
      {/* 3rd row */}
      {renderNumber(4)}
      {renderNumber(5)}
      {renderNumber(6)}
      {renderOprand('-', "orang")}
      {/* 4th row */}
      {renderNumber(1)}
      {renderNumber(2)}
      {renderNumber(3)}
      {renderOprand('+', "orang")}
      {/* 5th row */}
      {renderNumber(0, 'numberZero')}
      {renderNumber(10)}
      {renderOprand('=', "orang")}
    </div>
  );
}

export default Calculator;
