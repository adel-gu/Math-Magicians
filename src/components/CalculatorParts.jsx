import PropTypes from 'prop-types';
// Sub Components => {Screen, oprands, Numbers}
const Screen = ({ result }) => <span className="screen">{result}</span>;
Screen.propTypes = { result: PropTypes.string.isRequired };

// Oprands buttons components
const Oprand = ({ data, className, onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {data}
  </button>
);
Oprand.propTypes = { data: PropTypes.string.isRequired };
Oprand.propTypes = { className: PropTypes.string };
Oprand.propTypes = { onclick: PropTypes.func.isRequired };

// Numbers buttons components
const Number = ({ value, className , onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {value}
  </button>
);
Number.propTypes = { value: PropTypes.string.isRequired };
Number.propTypes = { className: PropTypes.string};
Number.propTypes = { onclick: PropTypes.func.isRequired};

export {Screen, Oprand, Number};