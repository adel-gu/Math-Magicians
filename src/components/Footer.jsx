const Footer = () => {
  const mb = {
    marginBottom: '10px',
  };
  return (
    <footer className="footer">
      <div className="container header-footer py-2" style={mb}>
        <span>
          Build by
          <a href="https://github.com/adel-gu">Adel Guitoun ❤️</a>
        </span>
      </div>
    </footer>
  );
};

export default Footer;
