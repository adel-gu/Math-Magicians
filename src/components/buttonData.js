const buttonsData = [
  { data: 'AC', columnPos: 1, rowPos: 2 },
  { data: '+/-', columnPos: 2, rowPos: 2 },
  { data: '%', columnPos: 3, rowPos: 2 },
  {
    data: '+',
    columnPos: 4,
    rowPos: 2,
    color: 'orang',
  },
  { data: '7', columnPos: 1, rowPos: 3 },
  { data: '8', columnPos: 2, rowPos: 3 },
  { data: '9', columnPos: 3, rowPos: 3 },
  {
    data: 'x',
    columnPos: 4,
    rowPos: 3,
    color: 'orang',
  },
  { data: '4', columnPos: 1, rowPos: 4 },
  { data: '5', columnPos: 2, rowPos: 4 },
  { data: '6', columnPos: 3, rowPos: 4 },
  {
    data: '-',
    columnPos: 4,
    rowPos: 4,
    color: 'orang',
  },
  { data: '1', columnPos: 1, rowPos: 5 },
  { data: '2', columnPos: 2, rowPos: 5 },
  { data: '3', columnPos: 3, rowPos: 5 },
  {
    data: '/',
    columnPos: 4,
    rowPos: 5,
    color: 'orang',
  },
  {
    data: '0',
    columnPos: 1,
    rowPos: 6,
    width: 2,
  },
  { data: '.', columnPos: 3, rowPos: 6 },
  {
    data: '=',
    columnPos: 4,
    rowPos: 6,
    color: 'orang',
  },
];

export default buttonsData;
