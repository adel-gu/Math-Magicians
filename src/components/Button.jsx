import React from 'react';
import PropTypes from 'prop-types';

// Stylesheets
import '../styles/Calculator.css';

export const Button = ({
  data,
  width,
  color,
  columnPos,
  rowPos,
  getBtnData,
}) => {
  const btnStyle = {
    gridColumn: `${columnPos} / span ${width}`,
    gridRow: rowPos,
  };
  return (
    <button
      style={btnStyle}
      className={`btn ${color}`}
      onClick={() => getBtnData(data)}
    >
      {data}
    </button>
  );
};

Button.defaultProps = {
  width: 1,
};

Button.propTypes = {
  data: PropTypes.string.isRequired,
  width: PropTypes.number,
  color: PropTypes.string,
  columnPos: PropTypes.number,
  rowPos: PropTypes.number,
  getBtnData: PropTypes.func.isRequired,
};
