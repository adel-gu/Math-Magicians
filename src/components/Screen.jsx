import React from 'react';
import PropTypes from 'prop-types';

// Stylesheets
import '../styles/Calculator.css';

export const Screen = ({ result }) => {
  return <span className="screen">{result}</span>;
};

Screen.defaultProps = {
  result: 0,
};

Screen.prototypes = {
  result: PropTypes.number,
};
