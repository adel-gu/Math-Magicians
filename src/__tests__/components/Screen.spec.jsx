import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Screen } from '../../components/Screen';

describe('Screen', () => {
  it('should render the screen', () => {
    render(<Screen result={0} />);

    const screenElement = screen.getByText(/0/i);

    expect(screenElement).toHaveTextContent('0');
    expect(screenElement).toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    const screenElement = renderer.create(<Screen result={0} />).toJSON();

    expect(screenElement).toMatchSnapshot();
  });
});
