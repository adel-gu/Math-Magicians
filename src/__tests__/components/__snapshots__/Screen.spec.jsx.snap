// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen should match the snapshot 1`] = `
<span
  className="screen"
  data-testid="screenTestId"
>
  0
</span>
`;
